taca_ngi_pipeline/deliver/deliver_dds.py

```diff
@@ -0,0 +1,546 @@
+            projectid,
+            sampleid,
+            **kwargs
+        )
+        self.config_statusdb = CONFIG.get('statusdb', None)
+        if self.config_statusdb is None and not do_release:
+            raise AttributeError("statusdb configuration is needed when delivering to DDS (url, username, password, port")
+        self.orderportal = CONFIG.get('order_portal', None)
+        if self.orderportal is None and not do_release:
+            raise AttributeError("Order portal configuration is needed when delivering to DDS")
+        if self.orderportal:
+            self._set_pi_details(pi_email, pi_name)
+            self._set_other_member_details(add_user, CONFIG.get('add_project_owner', False))
+            self._set_project_details(project_title, project_description)
+        self.sensitive = sensitive
+        self.fcid = fcid
+        
+    def get_delivery_status(self, dbentry=None):
+        """ Returns the delivery status for this project. If a dbentry
+        dict is supplied, it will be used instead of fethcing from database
+
+        :params dbentry: a database entry to use instead of
+        fetching from db
+        :returns: the delivery status of this project as a string
+        """
+        dbentry = dbentry or self.db_entry()
+        if dbentry.get('delivery_token'):
+            if dbentry.get('delivery_token') not in ['NO-TOKEN', 'not_under_delivery'] :
+                return 'IN_PROGRESS'  # At least some samples are under delivery
+        if dbentry.get('delivery_status'):
+            if dbentry.get('delivery_status') == 'DELIVERED':
+                return 'DELIVERED'  # The project has been marked as delivered
+        if dbentry.get('delivery_projects'):
+            return 'PARTIAL'  # The project underwent a delivery, but not for all the samples
+        return 'NOT_DELIVERED'  # The project is not delivered
+
+    def release_DDS_delivery_project(self, dds_project): 
+        """ Update charon when data upload is finished and release DDS project to user.
+        For this to work on runfolder deliveries, update the delivery status in Charon maually.
+        """
+        charon_status = self.get_delivery_status()
+        # Skip if status is not in progress
+        if charon_status != 'IN_PROGRESS':
+            logger.info("Project {} has no delivery token. Project is not being delivered at the moment".format(self.projectid))
+            return
+        question = "About to release project {} in DDS delivery project {} to user. Continue? ".format(self.projectid, dds_project)
+        if proceed_or_not(question):
+            logger.info("Releasing DDS project {} to user".format(dds_project))  
+        else:
+            logger.error("{} delivery has been aborted.".format(str(self)))
+            return
+        
+        delivery_status = 'IN_PROGRESS'
+        try:
+            cmd = ['dds', 'project', 'status', 'release', '--project', dds_project]
```

Comment:
Maybe add an optional deadline field so that we can control how long the delivery project is available if we want to.

Reply:
Good idea! I'll add that to my list of improvements 🙂